source/server/connection_handler_impl.cc

```diff
@@ -112,68 +121,57 @@ ConnectionHandlerImpl::getUdpListenerCallbacks(uint64_t listener_tag) {
 void ConnectionHandlerImpl::removeFilterChains(
     uint64_t listener_tag, const std::list<const Network::FilterChain*>& filter_chains,
     std::function<void()> completion) {
-  for (auto& listener : listeners_) {
-    if (listener.second.listener_->listenerTag() == listener_tag) {
-      listener.second.listener_->onFilterChainDraining(filter_chains);
-      break;
-    }
+  auto listener_it = listener_map_by_tag_.find(listener_tag);
+  if (listener_it != listener_map_by_tag_.end()) {
+    listener_it->second.second.listener_->onFilterChainDraining(filter_chains);
   }
+
   // Reach here if the target listener is found or the target listener was removed by a full
   // listener update. In either case, the completion must be deferred so that any active connection
   // referencing the filter chain can finish prior to deletion.
   Event::DeferredTaskUtil::deferredRun(dispatcher_, std::move(completion));
 }
 
 void ConnectionHandlerImpl::stopListeners(uint64_t listener_tag) {
-  for (auto& listener : listeners_) {
-    if (listener.second.listener_->listenerTag() == listener_tag) {
-      listener.second.listener_->shutdownListener();
-    }
+  auto iter = listener_map_by_tag_.find(listener_tag);
+  if (iter != listener_map_by_tag_.end()) {
+    iter->second.second.listener_->shutdownListener();
   }
 }
 
 void ConnectionHandlerImpl::stopListeners() {
-  for (auto& listener : listeners_) {
-    listener.second.listener_->shutdownListener();
+  for (auto& iter : listener_map_by_address_) {
```

Comment:
`listener_map_by_tag_` ?

Reply:
My final goal is to support multiple addresses listener. https://github.com/envoyproxy/envoy/issues/11184, 

The `listener_map_by_tag_`  will nest with a vector. For just iterate all the listeners, I just use the listener_map_by_address_ for more convenience, since it is just a flat map.

Reply:
If we finally complete it, will there are multiple address index for the same listener? Then we may call `shutdownListener()` method of same listener multiple times ?

In fact we can make the `listener_map_by_tag_` more simple just `absl::flat_hash_map<uint64_t, ActiveListenerDetailsPtr> listeners_`.

The reason why the address is saved in the previous `listeners_` list is to support searching by address. Now that we have a `listener_map_by_address_`, it is no longer necessary.

Reply:
> If we finally complete it, will there are multiple address index for the same listener? Then we may call `shutdownListener()` method of same listener multiple times?

The different address will have different listener. so it won't be called multiple times with shutdownListener().

But I think I still should use `listener_map_by_tag_`, since `listener_map_by_address_` only store the newest listener, the drained listener will be overridden at `addListener`. And the ListenerManager will stop the listener first, then add new listener, but I shouldn't depend on how `ListenerManager` invoke `ConnectionhandlerImpl`, that isn't part of API contract.

> 
> In fact we can make the `listener_map_by_tag_` more simple just `absl::flat_hash_map<uint64_t, ActiveListenerDetailsPtr> listeners_`.

Yea, That is much simple, I plan to add tag and address instance field into `ActiveListenerDetail`

> 
> The reason why the address is saved in the previous `listeners_` list is to support searching by address. Now that we have a `listener_map_by_address_`, it is no longer necessary.